src/api/c/memory.cpp

```diff
@@ -313,3 +317,439 @@ af_err af_get_mem_step_size(size_t *step_bytes) {
+    return *(MemoryManager *)handle;
+}
+
+af_memory_manager getHandle(MemoryManager &manager) {
+    MemoryManager *handle;
+    handle = &manager;
+    return (af_memory_manager)handle;
+}
+
+af_err af_create_memory_manager(af_memory_manager *manager) {
+    try {
+        AF_CHECK(af_init());
+        std::unique_ptr<MemoryManager> m(new MemoryManager());
+        *manager = getHandle(*m);
+        m.release();
+    }
+    CATCHALL;
+
+    return AF_SUCCESS;
+}
+
+af_err af_release_memory_manager(af_memory_manager handle) {
+    try {
+        // NB: does NOT reset the internal memory manager to be the default:
+        // af_unset_memory_manager_pinned must be used to fully-reset with a new
+        // AF default memory manager
+        delete (MemoryManager *)handle;
+    }
+    CATCHALL;
+
+    return AF_SUCCESS;
+}
+
+af_err af_set_memory_manager(af_memory_manager mgr) {
+    try {
+        std::unique_ptr<MemoryManagerFunctionWrapper> newManager(
+            new MemoryManagerFunctionWrapper(mgr));
+        // Calls shutdown() on the existing memory manager, but does not free
+        // the associated handle, if there is one
+        detail::setMemoryManager(std::move(newManager));
+    }
+    CATCHALL;
+
+    return AF_SUCCESS;
+}
+
+af_err af_unset_memory_manager() {
+    try {
+        detail::resetMemoryManager();
+    }
+    CATCHALL;
+
+    return AF_SUCCESS;
+}
+
+af_err af_set_memory_manager_pinned(af_memory_manager mgr) {
+    try {
+        // NB: does NOT free if a non-default implementation is set as the
+        // current memory manager - the user is responsible for freeing any
+        // controlled memory
+        std::unique_ptr<MemoryManagerFunctionWrapper> newManager(
+            new MemoryManagerFunctionWrapper(mgr));
+
+        // Calls shutdown() on the existing memory manager
+        detail::setMemoryManagerPinned(std::move(newManager));
+    }
+    CATCHALL;
+
+    return AF_SUCCESS;
+}
+
+af_err af_unset_memory_manager_pinned() {
+    try {
+        detail::resetMemoryManagerPinned();
+    }
+    CATCHALL;
+
+    return AF_SUCCESS;
+}
+
+af_err af_memory_manager_get_payload(af_memory_manager handle, void **payload) {
+    try {
+        MemoryManager &manager = getMemoryManager(handle);
+        *payload               = manager.payload;
+    }
+    CATCHALL;
+
+    return AF_SUCCESS;
+}
+
+af_err af_memory_manager_set_payload(af_memory_manager handle, void *payload) {
+    try {
+        MemoryManager &manager = getMemoryManager(handle);
+        manager.payload        = payload;
+    }
+    CATCHALL;
+
+    return AF_SUCCESS;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Native memory interface wrapper implementations
+
+af_err af_memory_manager_get_active_device_id(af_memory_manager handle,
+                                              int *id) {
+    try {
+        *id = memoryManager().getActiveDeviceId();
```

Comment:
This and the following functions should use the `af_memory_manager` handle passed into the function.

Reply:
Fixed in https://github.com/jacobkahn/arrayfire/commit/36c99b9d3a9d760782ff8edf4efd8fadebbb77ce